kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoWriterFactory.java

```diff
@@ -28,21 +32,37 @@ public class MongoWriterFactory<K> implements WriterFactory<K> {
 
   private final RemoteTable<K, WriteModel<Document>> table;
   private final MongoCollection<Document> genericCollection;
+  private final long epoch;
 
   public MongoWriterFactory(
       final RemoteTable<K, WriteModel<Document>> table,
-      final MongoCollection<Document> genericCollection
+      final MongoCollection<Document> genericCollection,
+      final MongoCollection<MetadataDoc> metadata,
+      final int kafkaPartition
   ) {
     this.table = table;
     this.genericCollection = genericCollection;
+    final MetadataDoc updateResult = metadata.findOneAndUpdate(
+        Filters.eq("partition", kafkaPartition),
+        Updates.inc("epoch", 1),
+        new FindOneAndUpdateOptions()
+            .upsert(true)
+            .returnDocument(ReturnDocument.AFTER)
+    );
+
+    if (updateResult == null) {
+      throw new IllegalStateException("Uninitialized metadata for partition " + kafkaPartition);
```

Comment:
shouldn't this be a task migrated exception?

Reply:
not sure why it would be that? this should never happen because the previous statement creates the document if it doesn't exist (there's no fencing that happens here)

Reply:
Ah right - not sure what I was thinking. Never mind.